internal/sci/kind/server.go

```diff
@@ -0,0 +1,95 @@
+package kind
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	sci "github.com/substratusai/substratus/internal/sci"
+)
+
+var _ sci.ControllerServer = &Server{}
+
+type Server struct {
+	SignedURLAddress string
+	BucketDir        string
+
+	sci.UnimplementedControllerServer
+}
+
+// ServeHTTP implements the http.Handler interface and is used to provide cloud-bucket-like signed URL support.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	log.Printf("Signed URL Server: %v", r.URL.Path)
+
+	switch r.Method {
+	case http.MethodPut:
+		// Expect "Content-Type: application/octet-stream" in a PUT request and save the body to a file.
+		if r.Header.Get("Content-Type") != "application/octet-stream" {
+			log.Print("client sent wrong content-type")
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		if err := s.saveUpload(r.Body, r.URL.Path); err != nil {
+			log.Printf("failed to save upload: %v", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+}
+
+func (s *Server) saveUpload(r io.Reader, urlPath string) error {
+	path := filepath.Join(s.BucketDir, urlPath)
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	if _, err := io.Copy(f, r); err != nil {
+		return err
+	}
```

Comment:
Reaching a bit here but is it better to `os.Stat(dst)` the destination before copying so as to not overwrite?

Reply:
I think overwriting the stored tarball is what we do in the real buckets today

Reply:
> overwriting the stored tarball is what we do in the real buckets today

Huh - I bet we have a not-so-subtle bug on GCP where one user could easily overwrite another's before the builder can start a build against it.

Reply:
Are you thinking two users could both be running `kubectl notebook -d .` on the same Notebook at the same time?

Reply:
If so, we are not currently accounting for that. Notebooks are intended to be used by 1 user at a time at the moment.